txpipe/twopoint.py

```diff
@@ -759,9 +792,9 @@ def calculate_shear_pos(self, i, j):
         else:
             rg = None
 
+        ng.calculateXi(rg=rg)
+        t2 = perf_counter()
         if self.rank == 0:
-            ng.calculateXi(rg=rg)
-            t2 = perf_counter()
             print(f"Processing took {t2 - t1:.1f} seconds")
 
         return ng
```

Comment:
seems to me as you are making each core calculate Xi now rather than only rank 0? Isn't Xi calculated based on the recombined results from all the processes? Is this intended?

Reply:
yes its intended. The issue was ng.xi is needed by all processes later in `write_output`. So if you only run calculateXi on rank==0 there will be an error on the other ranks. since calculateXi is so light it was easier to just have all processes run it

Reply:
alternatively we could compute ng.xi on rank==0 and bcast to the other ranks

Reply:
That is what confuses me cause if you look at line 414 in the write_ouput function other than rank==0 should terminate before they need the term estimated from calculateXi? I think at least that was the intend? @joezuntz any comments on this?

Reply:
I think the error comes before 414 in write output. in add_data_points

Here is the error 
```
Traceback (most recent call last):
 File "/opt/conda/lib/python3.9/runpy.py", line 197, in _run_module_as_main
Using new TreeCorr 4.3 or above
  return _run_code(code, main_globals, None,
 File "/opt/conda/lib/python3.9/runpy.py", line 87, in _run_code
  exec(code, run_globals)
 File "/global/cscratch1/sd/jelvinpo/DESC/TXPipe/txpipe/__main__.py", line 5, in <module>
  PipelineStage.main()
 File "/opt/conda/lib/python3.9/site-packages/ceci/stage.py", line 469, in main
  stage.execute(args)
 File "/opt/conda/lib/python3.9/site-packages/ceci/stage.py", line 604, in execute
  stage.run()
 File "/global/cscratch1/sd/jelvinpo/DESC/TXPipe/txpipe/twopoint.py", line 126, in run
  self.write_output(source_list, lens_list, meta, results)
 File "/global/cscratch1/sd/jelvinpo/DESC/TXPipe/txpipe/twopoint.py", line 419, in write_output
  self.add_data_points(S, results)
 File "/global/cscratch1/sd/jelvinpo/DESC/TXPipe/txpipe/twopoint.py", line 308, in add_data_points
  xi = d.object.xi
AttributeError: 'NNCorrelation' object has no attribute 'xi'
```

Reply:
The actual error comes from line 301

Reply:
You are correct it happens at 412 in `write_output`, when it calls `add_data_points`, question then from me is is there a reason that each cpu adds data points, or should that be initiated by just the rank==0? as far as I can see the `add_data_points` calls TreeCorr with a mpi comm object already. I think your solution solves the error, I will try and test it later. But it does add confusion for me to see how we are parallelizing these steps of the code. TreeCorr has built in mpi, which we are utilizing but are we also parallelizing the calls of TreeCorr themself?

Reply:
yea its a good point. only one rank really needs to construct the sacc file